From 5a8fdbf321e4e7dcd6a8c97c4a9d10e4eec75f4e Mon Sep 17 00:00:00 2001
From: ZhangXiao <xiao.zhang@windriver.com>
Date: Wed, 5 May 2010 12:44:37 +0800
Subject: [PATCH] sctp: IPsec rules are ineffective with ipv6

sctp: IPsec rules are ineffective with ipv6

For ipv6, sctp doesn't implement the IPsec mechanism thus it can't support
IPsec on IPv6. Enable the IPsec rules with sctp ipv6.

Signed-off-by: Zhang Xiao <xiao.zhang@windriver.com>
---
 net/sctp/ipv6.c |   16 ++++++++++++++++
 1 files changed, 16 insertions(+), 0 deletions(-)

diff --git a/net/sctp/ipv6.c b/net/sctp/ipv6.c
index 47f91af..0b044ec 100644
--- a/net/sctp/ipv6.c
+++ b/net/sctp/ipv6.c
@@ -245,6 +245,7 @@ static struct dst_entry *sctp_v6_get_dst(struct sctp_association *asoc,
 {
 	struct dst_entry *dst;
 	struct flowi fl;
+	int err;
 
 	memset(&fl, 0, sizeof(fl));
 	ipv6_addr_copy(&fl.fl6_dst, &daddr->v6.sin6_addr);
@@ -269,8 +270,23 @@ static struct dst_entry *sctp_v6_get_dst(struct sctp_association *asoc,
 		SCTP_DEBUG_PRINTK(
 			"rt6_dst:" NIP6_FMT " rt6_src:" NIP6_FMT "\n",
 			NIP6(rt->rt6i_dst.addr), NIP6(rt->rt6i_src.addr));
+		if (saddr)
+		{			
+			fl.proto = IPPROTO_SCTP;
+			if (asoc)
+				fl.oif = asoc->base.sk->sk_bound_dev_if;
+			fl.fl_ip_dport = daddr->v6.sin6_port;
+			fl.fl_ip_sport = saddr->v6.sin6_port;
+			if ((err = __xfrm_lookup(&dst, &fl, asoc ? asoc->base.sk : NULL, 0)) < 0) {
+				if (err == -EREMOTE)
+					err = ip6_dst_blackhole(asoc ? asoc->base.sk : NULL, &dst, &fl);
+				if (err < 0)
+					goto failure;
+			}
+		}	
 		return dst;
 	}
+failure:
 	SCTP_DEBUG_PRINTK("NO ROUTE\n");
 	dst_release(dst);
 	return NULL;
-- 
1.6.0.4

